tiledb/api/src/query/write/strategy.rs

```diff
@@ -0,0 +1,979 @@
+                        (c + 1) * approx_chunk_len
+                    };
+
+                    for r in self.records_included[0..exclude_min]
+                        .iter()
+                        .chain(self.records_included[exclude_max..].iter())
+                    {
+                        record_mask.set(*r)
+                    }
+                }
+
+                record_mask
+            }
+            Some(ShrinkSearchStep::Recur) | Some(ShrinkSearchStep::Done) => {
+                let mut record_mask = VarBitSet::new_bitset(self.nrecords);
+                for r in self.records_included.iter() {
+                    record_mask.set(*r);
+                }
+                record_mask
+            }
+        };
+
+        let fields = self
+            .field_mask
+            .iter()
+            .enumerate()
+            .filter(|(_, f)| f.is_included())
+            .map(|(i, _)| {
+                let f = self.schema.field(i);
+                (
+                    f.name.clone(),
+                    self.field_data[&f.name]
+                        .as_ref()
+                        .unwrap()
+                        .filter(&record_mask),
+                )
+            })
+            .collect::<Vec<(String, FieldData)>>();
+
+        WriteQueryData {
+            fields: fields.into_iter().collect(),
+        }
+    }
+
+    fn simplify(&mut self) -> bool {
+        self.explore_step(true)
+    }
+
+    fn complicate(&mut self) -> bool {
+        self.explore_step(false)
+    }
+}
+
+#[derive(Debug)]
+struct WriteQueryDataStrategy {
+    schema: Rc<SchemaData>,
+}
+
+impl WriteQueryDataStrategy {
+    pub fn new(schema: &Rc<SchemaData>) -> Self {
+        WriteQueryDataStrategy {
+            schema: Rc::clone(schema),
+        }
+    }
+}
+
+impl Strategy for WriteQueryDataStrategy {
+    type Tree = WriteQueryDataValueTree;
+    type Value = WriteQueryData;
+
+    fn new_tree(&self, runner: &mut TestRunner) -> NewTree<Self> {
+        const WRITE_QUERY_MIN_RECORDS: usize = 0;
+        const WRITE_QUERY_MAX_RECORDS: usize = 1024 * 1024;
+
+        const WRITE_QUERY_MIN_VAR_SIZE: usize = 0;
+        const WRITE_QUERY_MAX_VAR_SIZE: usize = 1024 * 128;
+
+        /* Choose the maximum number of records */
+        let nrecords = (WRITE_QUERY_MIN_RECORDS..=WRITE_QUERY_MAX_RECORDS)
+            .new_tree(runner)?
+            .current();
+
+        /* generate a random set of fields to query */
+        let field_mask = {
+            let ndimensions = self.schema.domain.dimension.len();
+            let nattributes = self.schema.attributes.len();
+
+            let dimensions_mask = match self.schema.array_type {
+                ArrayType::Dense => {
+                    /* dense array coordinates are handled by a subarray */
+                    vec![WriteFieldMask::Exclude; ndimensions]
+                }
+                ArrayType::Sparse => {
+                    /* sparse array must write coordinates */
```

Comment:
The strategy has to go further than this and restrict the range of what is written to the sparse dimensions. That's a big TODO in order to get this to pass